No descriptor ids in spk txout index

LLFourn opened the pull request:
Fixes #1459

This reverts part of the changes in #1203. There the `SpkTxOutIndex<(K,u32)>` was changed to `SpkTxOutIndex<(DescriptorId, u32>)`. This led to a complicated translation logic in  `KeychainTxOutIndex` (where the API is based on `K`) to transform calls to it to calls to the underlying `SpkTxOutIndex` (which now indexes by `DescriptorId`). The translation layer was broken when it came to translating range queries from the `KeychainTxOutIndex`. My solution was just to revert this part of the change and remove the need for a translation layer (almost) altogether. A thin translation layer remains to ensure that un-revealed spks are filtered out before being returned from the `KeychainTxOutIndex` methods.

I feel like this PR could be extended to include a bunch of ergonomics improvements that are easier to implement now. But I think that's the point of https://github.com/bitcoindevkit/bdk/pull/1451 so I held off and should probably go and scope creep that one instead.


### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing


#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

LLFourn committed:
Write failing test for keychain range querying

LLFourn committed:
refactor(keychain): Fix KeychainTxOutIndex range queries

The underlying SpkTxOutIndex should not use DescriptorIds to index
because this loses the ordering relationship of the spks so queries on
subranges of keychains work.

Along with that we enforce that there is a strict 1-to-1 relationship
between descriptors and keychains. Violating this leads to an error in
insert_descriptor now.

In general I try to make the translation layer between the SpkTxOutIndex
and the KeychainTxOutIndex thinner. Ergonomics of this will be improved
in next commit.

The test from the previous commit passes.

LLFourn committed:
chore(chain): misc docs and insert_descriptor fixes

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -50,9 +51,17 @@ impl<K: Ord> Append for ChangeSet<K> {
+            {
+                self.keychains_added.insert(new_keychain, new_descriptor);
+            }
+        }
```
Can we have a simple comment explaining why we are doing this?
I.e. We disallow re-assigning a keychain's descriptor or assigning the same descriptor to multiple keychains.

LLFourn replied:
I added more docs and comment on this: 558a33a205d80c0e4bb3d3a24bdc7b826a90c8e0

notmandatory committed:
chore(chain): add type IndexSpk, fix clippy type complexity warning

notmandatory commented on clippy.toml:
```diff
@@ -1 +1,2 @@
 msrv="1.63.0"
+type-complexity-threshold = 275
```
🔧 Since our types really are too complex can we just add a new `IndexSpk` type alias to fix the problem instead? It would also improve the docs for other APIs in the modules that also use `(u32,ScriptBuf)`. See my proposed patch here (feel free to cherry-pick or copy paste): https://github.com/notmandatory/bdk/commit/e15565571dd2d55d54c5d31de1042bbcc381daf4

We don't need to use the type outside of the `bdk_chain` crate but it shouldn't hurt since the other crates will always depend on chain.

LLFourn replied:
I made another commit to take this idea further: 4eb1e288a9362803b034590e0c56c8cf9cf8b0c2

There's still a bit to be done to have good ergonomics but I think that's better in followup.

I made the spks owned like in #1341

notmandatory commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -50,9 +51,17 @@ impl<K: Ord> Append for ChangeSet<K> {
+                    .keychains_added
+                    .values()
+                    .all(|descriptor| descriptor != &new_descriptor)
+            {
```
🔧 Would this work as a more efficient single pass, short-circuiting way to do this?
```suggestion
            if self.keychains_added.iter().all(|(keychain, descriptor)| {
                keychain != &new_keychain && descriptor != &new_descriptor
            }) {
```

LLFourn replied:
No this code looks wrong. It's meant to check that no descriptor already in the changeset is the same as the one we're adding. The correct way to do this is a bijective map struct i.e. a BTreeMap one way and HashMap the other way. This might be handy to have in general but I guess it belongs in another PR.

notmandatory replied:
My suggested code does check that neither the new `keychain` or `descriptor` being added exists in the set of `keychains_added` just as the original code, but should be O(n) instead of O(2n). 

But I'm ok leaving this as is for now and worrying about the efficiency later if it ever becomes a problem.

LLFourn replied:
Sorry I understand what you are saying now. `contains_key` is O(log(n)) so keeping the fast one fast and separate seemed sensible.

notmandatory commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -112,22 +122,24 @@ const DEFAULT_LOOKAHEAD: u32 = 25;
 ///
-/// Methods that can update the last revealed index or add keychains will return [`super::ChangeSet`] to report
+/// Methods that can update the last revealed index or add keychains will return [`ChangeSet`] to report
 /// these changes. This can be persisted for future recovery.
```
⛏️  Would be good to make this recommendation stronger here.
```suggestion
/// these changes. This [`ChangeSet`] should be persisted for future recovery.
```

notmandatory commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -206,22 +190,18 @@ const DEFAULT_LOOKAHEAD: u32 = 25;
 pub struct KeychainTxOutIndex<K> {
-    inner: SpkTxOutIndex<(DescriptorId, u32)>,
+    inner: SpkTxOutIndex<(K, u32)>,
     // keychain -> (descriptor, descriptor id) map
```
🔧  needs to match the code change.
```suggestion
    // keychain -> descriptor id map
```

notmandatory commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -462,66 +408,82 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
-    /// overwritten.
+    /// keychain <-> descriptor is a one-to-one mapping -- in `--release` this method will ignore calls that try to
+    /// associate a keychain with the descriptor of another keychain or to re-assign the keychain to
+    /// new descriptor. If `debug_assertions` are enabled then it will panic.
```
🔧 This code now throws an error and should not panic, docs need to be updated to reflect new behavior.
```suggestion
    /// The keychain <-> descriptor relationship is a one-to-one mapping that can not be changed. This method 
    /// will throw an error if you try to associate a new keychain with an already assigned descriptor or to re-assign 
    /// an existing keychain to a new descriptor.
```

notmandatory reviewed (changes requested):
LGTM but I have a few suggestions before ACKing.

LLFourn committed:
chore(chain): Standardise KeychainTxOutIndex return types

The previous commit b9c5b9d08b040faf6c6b2d9b3745918031555b72 added
IndexSpk. This goes further and adds `Indexed` and `KeychainIndexed`
type alises (IndexSpk is Indexed<ScriptBuf>) and attempts to standardize
the structure of return types more generally.

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -233,36 +214,39 @@ impl<K> Default for KeychainTxOutIndex<K> {
                 .map(|(k, v)| (k.clone(), v.clone()))
                 .collect(),
-            last_revealed: self.last_revealed.clone(),
+            last_revealed: self.last_revealed.clone().into_iter().collect(),
```
Nit:

```suggestion
            last_revealed: self.last_revealed.iter().map(|(did, i)| (*did, *i)).collect(),
```

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -462,66 +418,82 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+            changeset
+                .keychains_added
+                .insert(keychain.clone(), descriptor);
+        } else {
```
Nit: else branch not needed.

LLFourn replied:
why is it not needed?

evanlinjin replied:
Oh it is needed sorry.

LLFourn committed:
chore(chain): Improve documentation of keychain::ChangeSet

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -548,114 +444,146 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+        if let Some(keychain) = self.descriptor_ids_to_keychains.get(&did).cloned() {
+            self.replenish_lookahead(&keychain, self.lookahead);
+        }
+    }
```
I don't think this method is necessary. We are only using it in 2 places; `Indexer::index_txout` and `apply_changeset`.

For `index_txout`, the keychain is already available for us (however we may need to clone the `inner.scan_txout` result because of the referencing rules of rust).

For `apply_changeset`, we can just query `Self::descriptor_ids_to_keychains` within the loop.

LLFourn replied:
I think the method is useful to explicitly make the choice to silently ignore the corresponding `did` or keychain not existing. Doing the lookups in place means me or someone else is likely to just do `.expect("invariant")` in that case without understanding context and reviewers are unlikely to notice it.

I take your point that in one place the keychain already exists. I refactored the methods to take advantage of that.

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -548,114 +444,146 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+            .all_spks()
+            .range(self.map_to_inner_bounds((start, end)));
+        let mut current_keychain = iter_last_revealed.next();
+        // The reason we need a tricky algorithm is because of the "lookahead" feature which means
+        // that some of the spks in the SpkTxoutIndex will not have been revealed yet. So we need to
+        // filter out those spks that are above the last_revealed for that keychain. To do this we
+        // iterate through the last_revealed for each keychain and the spks for each keychain in
+        // tandem. This minimizes BTreeMap queries.
+        core::iter::from_fn(move || loop {
+            let ((keychain, index), spk) = iter_spks.next()?;
+            // We need to find the last revealed that matches the current spk we are considering so
+            // we skip ahead.
+            while current_keychain?.0 < keychain {
+                current_keychain = iter_last_revealed.next();
+            }
+            let (current_keychain, last_revealed) = current_keychain?;
+
+            if current_keychain == keychain && Some(*index) <= last_revealed {
+                break Some(((keychain.clone(), *index), spk.as_script()));
+            }
+        })
```
I'm trying to understand the performance implications of this (since this complex-ish logic seems to be here for performance reasons, otherwise a `.filter` can work too).

I.e. let's say the caller is using a large lookahead (1000+) with a large amount of keychains. Would it be more performant to loop through all lookahead spks (except for the last keychain) - as we are doing here. Or, would it be more performant to do multiple `.range` calls on `inner.all_spks()` per keychain?

My argument is that performance is more of an issue for usecases with larger keychains-counts/spk-counts/lookahead-values.

LLFourn replied:
Indeed this is written so it hits the optimal time complexity in the number of keychains.

Let $k$ is the number of keychains total, $s$ is the average number of spks per keychain and $h$ is the lookahead value. 

Let's consider the worst case performance when the query matches all keychains and $h = 0$. For the code above, there are two map queries which take negligible time and then we iterate through all the spks of every matched keychain so the time is dominated by the $sk$ spks you iterate through.

For the previous implementation where you do two lookups you have to do $sk$ iterations but you have another term that could take more time: $k\log(sk)$ time to do the btree map lookups which domainates when $\log(sk) > s$. The bottom line is that keeping everything other than $k$ constant the complexity is of PR impl is $O(k)$ while the previous impl was $O(k\log(k))$ so it's a significant improvement in theory.

Setting $h > 0$ it becomes $k(s+h)$ (this PR) vs $k\log(k(s+h))$ (prev). In the case that $h$ dominates $s$ which is maximally advantages the former implementation you get $kh$ vs $k(\log(k) + \log(h))$. Therefore the prev impl would always be faster where the lookahead dominates the actual number of spks. To improve performance for applications where $h >> s$ I'd say we should simply add a revleaed spks cache BTreeMap.

evanlinjin reviewed (commented):
Thanks for working on this. The range-based method bugs definitely need to be fixed. However, I'm not sure if a total-revamp of the behavior is the correct way forward?

I'm okay with disallowing reassignment of a keychain's descriptor. However, I still think assigning the same descriptor to multiple keychains is important (for consistency). Let me explain.

Currently, our API allows adding *overlapping* descriptors. I.e. a non-wildcard descriptor which derives a spk that exists in a wildcard descriptor. There is just no good way to detect for this. With the changes presented in this PR, doing the above (which is impossible to detect) will cause undefined behavior, whereas the previous implementation handles this perfectly.

Therefore, for consistency, I think we should just enable both and scenarios and handle things consistently (as we are doing before).

evanlinjin commented:
However, I really like the API changes done in this PR. I.e.
* Returning a `Vec` of revealed spks (instead of iterator)
* Introducing `KeychainIndexed` and `Indexed` type parameters (however, can we think of a better name?)

ValuedMammal commented:
> There is just no good way to detect for this.

The way to detect it is to check the value returned by `self.inner.insert_spk()` instead of ignoring it.

ValuedMammal commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -803,50 +665,67 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+            }
+        }
+
+        Some((spks, changeset))
```
The way it's written this will never return `None`, I think?

evanlinjin replied:
What if the keychain does not exist?

LLFourn replied:
I think I might try and make an Error for the keychain not exisitng.

LLFourn replied:
(nope it is fine as it is -- and evan is right it will return `None` when there is not keychain).

ValuedMammal replied:
You could add a line to check `self.keychain_to_descriptor_id`

```diff
     pub fn reveal_to_target(
         &mut self,
         keychain: &K,
         target_index: u32,
     ) -> Option<(Vec<Indexed<ScriptBuf>>, ChangeSet<K>)> {
+        let _ = self.keychain_to_descriptor_id.get(keychain)?;
         let mut changeset = ChangeSet::default();
+
         let mut spks: Vec<Indexed<ScriptBuf>> = vec![];
         while let Some((i, new)) = self.next_index(keychain) {
             if !new || i > target_index {
                 break;
             }
             match self.reveal_next_spk(keychain) {
                 Some(((i, spk), change)) => {
                     spks.push((i, spk));
                     changeset.append(change);
                 }
                 None => break,
             }
         }
 
         Some((spks, changeset))
     }
```

ValuedMammal commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -803,50 +665,67 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+            let _ = self.inner.insert_spk((keychain.clone(), next_index), spk);
+            self.last_revealed.insert(*did, next_index);
+            changeset.last_revealed.insert(*did, next_index);
+            self.replenish_lookahead(keychain, self.lookahead);
```
I'm thinking as long as we update `self.last_revealed` and then call `replenish_lookahead`, we shouldn't have to insert the spk here.

LLFourn replied:
Ah nice catch. Hmm I would then change the name of the `lookahead` functions so that they are not just for lookaheads.

see 1b6fe8cb0ae0104dac81730e34972804c04114b5

evanlinjin commented:
> > There is just no good way to detect for this.
> 
> The way to detect it is to check the value returned by `self.inner.insert_spk()` instead of ignoring it.

Yes, but that means we need to error when that happens (since it's non-viable to iterate through all of the wildcard-descriptor's spks). This means all methods that may need to derive more spks have to return an error. I don't think this is a reasonable API.

ValuedMammal commented on crates/chain/src/keychain.rs:
```diff
@@ -49,6 +49,11 @@ impl Balance {
 }
 
+/// A tuple of keychain index and corresponding [`ScriptBuf`].
+pub type Indexed<T> = (u32, T);
```
maybe something like

```suggestion
/// A tuple of keychain index and an arbitrary `T` representing the indexed value.
pub type Indexed<T> = (u32, T);
```

LLFourn committed:
chore(chain): Fix Indexed and KeychainIndexed documentaion

Co-authored-by: ValuedMammal <valuedmammal@protonmail.com>

LLFourn committed:
refactor(chain): improve replenish lookeahd internals

see: https://github.com/bitcoindevkit/bdk/pull/1463/files/4eb1e288a9362803b034590e0c56c8cf9cf8b0c2#r1630943639

LLFourn committed:
chore(chain): document insert_descriptor invariants better

LLFourn commented:
> With the changes presented in this PR, doing the above (which is impossible to detect) will cause undefined behavior

The two claims made here are wrong in my view:

1. **impossible to detect**: Figuring out if two descriptors can produce the same spk is not a hard problem of computer science. I bet you can come up with a pretty good algorithm and implementation for this in less than a day. You can do a decent job of it by just by replacing the definite derivation indices with wildcards of any descriptor your adding and using that as a key in a lookup. I really don't think you should do this -- it's not worth your time because checking whether they're equal and documenting that the thing should not be used that way is good enough. This is just the place to solve the problem. Making the internals here complicated is certainly not.
2. **undefined behavior**: The first first keychain to derive that descriptor will be the owner of it. To me this seems like a simple, good and well defined behavior. Perhaps you mean it hasn't been defined in terms of writing a test. Feel free to do so.

update: I improved docs around this in d2629cbc320192e846d4c6dd9b56c7cc18764d3c also mentioning that it's up to the user not to insert overlapping descriptors.

LLFourn committed:
chore(chain): s/replenish_lookahead/replenish_inner_index/

evanlinjin commented:
I've been thinking about `KeychainTxOutIndex` overnight, and the direction that this PR takes us is growing on me.

The keychain `K` is useful for the lifetime of `KeychainTxOutIndex` so that we can do range queries using `K` or `(K, u32)` for spks & indexed outputs.

For persisting, it's a more accurate representation to persist the `last_revealed_index` with the corresponding descriptor (instead of the keychain `K`). It's also handy if the user wishes to reconstruct the `KeychainTxOutIndex` later with a different set of keychains.

In addition, I was in discussion with @LLFourn yesterday. He suggested that we should consider NOT persisting descriptors via the `keychain::ChangeSet` since some users do not care about persisting the descriptor here.

evanlinjin committed:
refactor(chain): `KeychainTxOutIndex` use `HashMap` for fields

Instead of `BTreeMap` which is less performant.

evanlinjin commented:
Steps after this PR (not in any specific order):
* Rename generic `K` (keychain) to `L` (label). Label represents how we want to "label" the keychains for the lifetime of `KeychainTxOutIndex`.
* Introduce `trait Keychain` that produces `KeychainId` (which replaces `DescriptorId`) and a method that produces spks. A keychain always produces the same spks.
* `KeychainId` is created from a hash of the concatenation of 0th derived spk and a flag for wildcard/non-wildcard. This makes the `KeychainId` better for uniquely-identifying a keychain (than the hash of the miniscript descriptor).

evanlinjin commented:
I'm tempted to remove all instances where we prematurely collect, but can leave it for another PR.

evanlinjin committed:
docs(chain): fix docs

evanlinjin committed:
refactor(chain): change field names to be more sane

evanlinjin reviewed (approved):
self-ACK 3070b2dfaf7b48e4280dc2c5805fa76b5058245c

ValuedMammal reviewed (commented):
ACK 3070b2dfaf7b48e4280dc2c5805fa76b5058245c

plus an additional comment about `reveal_to_target`

notmandatory reviewed (approved):
tACK 3070b2dfaf7b48e4280dc2c5805fa76b5058245c

I also manually tested the examples with signet descriptors and got back no errors and expected balance after scan and sync.

evanlinjin committed:
refactor(chain): compute txid once for `KeychainTxOutIndex::index_tx`

evanlinjin reviewed (approved):
ACK 8dd174479f9719309663ed979a5b4b86aca0a6e9
